presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerResourceGroupService.java

```diff
@@ -0,0 +1,87 @@
+import com.facebook.presto.execution.resourceGroups.ResourceGroupRuntimeInfo;
+import com.facebook.presto.metadata.InternalNode;
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import io.airlift.units.Duration;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.cache.CacheLoader.asyncReloading;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class ResourceManagerResourceGroupService
+        implements ResourceGroupService
+{
+    private final DriftClient<ResourceManagerClient> resourceManagerClient;
+    private final InternalNodeManager internalNodeManager;
+    private final Cache<InternalNode, List<ResourceGroupRuntimeInfo>> cache;
+    private final Executor executor = Executors.newCachedThreadPool();
+
+    @Inject
+    public ResourceManagerResourceGroupService(
+            @ForResourceManager DriftClient<ResourceManagerClient> resourceManagerClient,
+            ResourceManagerConfig resourceManagerConfig,
+            InternalNodeManager internalNodeManager)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerService is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        Duration cacheExpireDuration = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null").getResourceGroupServiceCacheExpireInterval();
+        Duration cacheRefreshDuration = resourceManagerConfig.getResourceGroupServiceCacheRefreshInterval();
+        this.cache = CacheBuilder.newBuilder()
+                .expireAfterWrite(cacheExpireDuration.roundTo(MILLISECONDS), MILLISECONDS)
+                .refreshAfterWrite(cacheRefreshDuration.roundTo(MILLISECONDS), MILLISECONDS)
+                .build(asyncReloading(new CacheLoader<InternalNode, List<ResourceGroupRuntimeInfo>>() {
+                    @Override
+                    public List<ResourceGroupRuntimeInfo> load(InternalNode internalNode)
+                            throws ResourceManagerInconsistentException
+                    {
+                        return getResourceGroupInfos(internalNode);
+                    }
+                }, executor));
+    }
+
+    @Override
+    public List<ResourceGroupRuntimeInfo> getResourceGroupInfo()
+            throws ResourceManagerInconsistentException
+    {
+        try {
+            InternalNode currentNode = internalNodeManager.getCurrentNode();
+            return cache.get(currentNode, () -> getResourceGroupInfos(currentNode));
+        }
+        catch (ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)